google-cloud-debugger/docs/design.md

```diff
@@ -0,0 +1,366 @@
+# Stackdriver Debugger Agent Design Doc
+
+The `google-cloud-debugger` instruments Stackdriver Debugger service in Ruby 
+applications by implementing a Debugger Agent, which is able to communicate
+with backend service to fetch user defined breakpoints and facilitate the
+breakpoints evaluation in running Ruby applications. This document explains the 
+design of the agent, how it works, and some rationale behind the design 
+decisions.
+
+## General workflow
+
+A typical Stackdriver Debugger workflow starts with a debugger agent registers
+itself and the application upon application startup. The registered debuggee
+application/service will become avaiable on the [Stackdriver
+ Debugger console](http://console.cloud.google.com/debug). Then user can select
+application source code, if not automatically selected, and create breakpoints
+on the UI. These breakpoints will be automatically distributed to all the 
+debugger agents registered under this same service. The agents then turn on the
+built-in tracing machanism to monitor the running applications. If any
+application instance triggers the code with breakpoint defined, the breakpoints 
+will be evaluated without affecting the running application, and the data 
+will be sent back to the backend service and displayed on the Debugger UI. 
+When all the breakpoints are evaluated or expired, the debugger agent will 
+deactivate the program tracing.
+
+## Architecture
+
+The Stackdriver Debugger agent uses two child threads to handle communication 
+with backend Stackdriver Debugger service. One thread for long polling 
+breakpoints and one thread for submitting evaluated breakpoint data. The local 
+breakpoints are managed by a shared Breakpoint Manager object and they are 
+traced by using using Ruby VM Tracing API at C level.
+
+![google-cloud-debugger agent 
+design](https://gist.githubusercontent.com/hxiong388/9fdfec068a218f83cc9bfc7fe6ce587d/raw/5810920af8e9161445c9198adcf18a063077457b/google-cloud-debugger-agent-design.svg?sanitize=true
```

Comment:
Can you check these SVGs into the docs directory rather than having them in your personal gists (where no one else can edit them)?

Reply:
Fixed